pkg/v1/remote/options.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"net/http"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+)
+
+// WithTransport is a functional option for overriding the default transport
+// on a remote image
+func WithTransport(t http.RoundTripper) ImageOption {
+	return func(i *ImageOpener) error {
+		return i.SetTransport(t)
+	}
+}
+
+// WithAuth is a functional option for overriding the default authenticator
+// on a remote image
+func WithAuth(auth authn.Authenticator) ImageOption {
+	return func(i *ImageOpener) error {
+		return i.SetAuth(auth)
+	}
+}
+
+// Set client on image using provided transport, and the default authenticator
+func (i *ImageOpener) SetTransport(t http.RoundTripper) error {
+	auth, err := authn.DefaultKeychain.Resolve(i.ref.Context().Registry)
+	if err != nil {
+		return err
+	}
+	tr, err := transport.New(i.ref.Context().Registry, auth, t, transport.DefaultScopes(i.ref))
+	if err != nil {
+		return err
+	}
+	i.client = &http.Client{Transport: tr}
+	return nil
```

Comment:
These should be stored and lazily evaluated.  I think by evaluating these eagerly there are assumptions and bugs.

Reply:
There are other side effects of doing this eagerly, like multiple registry roundtrips, multiple cred helper invocations, ...

Reply:
whoops, yeah these are already being evaluated in `Open()`. meant to remove this